In [71]:
import numpy as np
import numpy.linalg as LA

In [72]:
X = np.array([[1, 0.48, 1.52,0.71],
              [2.05, 1, 3.26, 1.56],
              [0.64, 0.3, 1, 0.46],
              [1.41, 0.61, 2.08, 1]]
            )

In [73]:
X[1][2]

3.26

In [74]:
LA.svd(X).S

array([5.63015402, 0.0818276 , 0.02969673, 0.01371211])

Max is 12.88

In [75]:
X@X@X@X@X

array([[246.13052011, 114.91353727, 378.04729761, 178.26899082],
       [521.27549005, 243.37335541, 800.65970886, 377.55275394],
       [158.1363764 ,  73.83078854, 242.89157537, 114.53602844],
       [335.63770644, 156.702696  , 515.52699737, 243.09782984]])

In [76]:
X

array([[1.  , 0.48, 1.52, 0.71],
       [2.05, 1.  , 3.26, 1.56],
       [0.64, 0.3 , 1.  , 0.46],
       [1.41, 0.61, 2.08, 1.  ]])

In [77]:
np.tensordot(X,X)

array(31.7064)

In [78]:
a = np.arange(60.).reshape(3,4,5)
b = np.arange(24.).reshape(4,3,2)
c = np.tensordot(a,b, axes=([1,0],[0,1]))

In [79]:
a

array([[[ 0.,  1.,  2.,  3.,  4.],
        [ 5.,  6.,  7.,  8.,  9.],
        [10., 11., 12., 13., 14.],
        [15., 16., 17., 18., 19.]],

       [[20., 21., 22., 23., 24.],
        [25., 26., 27., 28., 29.],
        [30., 31., 32., 33., 34.],
        [35., 36., 37., 38., 39.]],

       [[40., 41., 42., 43., 44.],
        [45., 46., 47., 48., 49.],
        [50., 51., 52., 53., 54.],
        [55., 56., 57., 58., 59.]]])

In [80]:
X

array([[1.  , 0.48, 1.52, 0.71],
       [2.05, 1.  , 3.26, 1.56],
       [0.64, 0.3 , 1.  , 0.46],
       [1.41, 0.61, 2.08, 1.  ]])

In [81]:
np.tensordot(X, X, axes = 0)

array([[[[ 1.    ,  0.48  ,  1.52  ,  0.71  ],
         [ 2.05  ,  1.    ,  3.26  ,  1.56  ],
         [ 0.64  ,  0.3   ,  1.    ,  0.46  ],
         [ 1.41  ,  0.61  ,  2.08  ,  1.    ]],

        [[ 0.48  ,  0.2304,  0.7296,  0.3408],
         [ 0.984 ,  0.48  ,  1.5648,  0.7488],
         [ 0.3072,  0.144 ,  0.48  ,  0.2208],
         [ 0.6768,  0.2928,  0.9984,  0.48  ]],

        [[ 1.52  ,  0.7296,  2.3104,  1.0792],
         [ 3.116 ,  1.52  ,  4.9552,  2.3712],
         [ 0.9728,  0.456 ,  1.52  ,  0.6992],
         [ 2.1432,  0.9272,  3.1616,  1.52  ]],

        [[ 0.71  ,  0.3408,  1.0792,  0.5041],
         [ 1.4555,  0.71  ,  2.3146,  1.1076],
         [ 0.4544,  0.213 ,  0.71  ,  0.3266],
         [ 1.0011,  0.4331,  1.4768,  0.71  ]]],


       [[[ 2.05  ,  0.984 ,  3.116 ,  1.4555],
         [ 4.2025,  2.05  ,  6.683 ,  3.198 ],
         [ 1.312 ,  0.615 ,  2.05  ,  0.943 ],
         [ 2.8905,  1.2505,  4.264 ,  2.05  ]],

        [[ 1.    ,  0.48  ,  1.52  ,  0.71  ],
 

Three common use cases are:

axes = 0 : tensor product

axes = 1 : tensor dot product

axes = 2 : (default) tensor double contraction

In [83]:
X

array([[1.  , 0.48, 1.52, 0.71],
       [2.05, 1.  , 3.26, 1.56],
       [0.64, 0.3 , 1.  , 0.46],
       [1.41, 0.61, 2.08, 1.  ]])

In [82]:
max = 1

for i in range(4):
        print(X[3][i] * X[i][3])


1.0010999999999999
0.9516
0.9568000000000001
1.0


In [84]:
1.41*0.71 # instant profit to be made off the bat!

1.0010999999999999

In [87]:
((1.41*0.71)**2 - 1)*100 # at minumum we can make 0.22% but we can definetly go higher than this

0.22012099999997314

In [89]:
max = 1

for i in range(4):
        for j in range(4):
                print(X[3][i] * X[i][j] * X[j][3])



1.0010999999999999
1.055808
0.985872
1.0010999999999999
0.887855
0.9516
0.914756
0.9516
0.9451520000000001
0.9734400000000001
0.9568000000000001
0.9568000000000001
1.0010999999999999
0.9516
0.9568000000000001
1.0


In [94]:
max = 1
a = 0
b = 0
c = 0
d = 0

for i in range(4):
    for j in range(4):
        for k in range(4):
            for l in range(4):
                res = X[3][i] * X[i][j] * X[j][k] * X[k][l] * X[l][3]
                if res > max: 
                    print(X[3][i] * X[i][j] * X[j][k] * X[k][l] * X[l][3])
                    max = res
                    a = i
                    b = j
                    c = k
                    d = l

1.0010999999999999
1.055808
1.0569693888


In [95]:
print(a, b, c, d)

0 1 3 0


In [102]:
print(X[3][a] * X[a][b] * X[b][c] * X[c][d] * X[d][3])


1.0569693888


In [106]:
X[b][c]

1.56

In [108]:
print(X[3][a] * X[a][b]) # pizza slice to wasabi root
X[3][a] * X[a][b] * X[b][c] # wasabi root to shells

0.6768


1.055808

In [1]:
1.055808 * 1.41

1.48868928

According to the numbers above, the best trade is as follows:

1. Trade 1 unit of shells for 1.41 units of pizza slices
2. Trade 1.41 units of pizza slices for 0.6768 units of wasabi root
3. Trade 0.6768 units of wasabi root for 1.055808 units of shells
4. Trade 1.055808 units of shells for 1.48868928 units of pizza sclies
5. Trade back to shells for a shell value of of **1.0569693888** 

This equates roughly to a 5.7% rate of return off of 5 trades